Tests/PostgresNIOTests/New/PostgresConnectionTests.swift

```diff
@@ -638,6 +638,63 @@ class PostgresConnectionTests: XCTestCase {
         }
     }
 
+    func testQueriesFailIfConnectionIsClosed() async throws {
+        let (connection, channel) = try await self.makeTestConnectionWithAsyncTestingChannel()
+
+        try await connection.closeGracefully()
+
+        XCTAssertEqual(channel.isActive, false)
+
+        do {
+            _ = try await connection.query("SELECT version;", logger: self.logger)
+            XCTFail("Expected to fail")
+        } catch let error as ChannelError {
+            XCTAssertEqual(error, .ioOnClosedChannel)
+        }
+    }
+
+    func testListenFailsIfConnectionIsClosed() async throws {
+        let (connection, channel) = try await self.makeTestConnectionWithAsyncTestingChannel()
+
+        try await connection.closeGracefully()
+
+        XCTAssertEqual(channel.isActive, false)
+
+        do {
+            _ = try await connection.listen("test_channel")
+            XCTFail("Expected to fail")
+        } catch let error as ChannelError {
+            XCTAssertEqual(error, .ioOnClosedChannel)
+        }
+    }
+
+    func testListenFailsIfConnectionIsClosedMidway() async throws {
```

Comment:
What connection state do you want to test here? In which connection state does this happen? Can we mock this explicitly and not by accident?

Reply:
The test is consistent, it only waits 3 seconds to make sure the listener has started listening.

It tests that if there is a listen in process, and then the connection is closed, the listener sequence needs to fail.

Reply:
ah! in that case we should wait for the client messages on the channel instead of using a timer.

See here how it is done:
https://github.com/vapor/postgres-nio/blob/7b621c16f6a0a8a0af8badd56b6f980457a1b7c5/Tests/PostgresNIOTests/New/PostgresConnectionTests.swift#L96

Reply:
Thanks for the clues, I did this in the new test.
We no longer need to change that `fatalError("Invalid state: \(self.state)")`, it appears that was triggered as a result of my bad test.

@fabianfett i think i no longer need to make another PR out of this?

Reply:
@fabianfett I think this PR's changes are good, but I noticed this results in a fatal error regardless of my changes:
```swift
    func testAddListenerFailsIfConnectionIsClosed() async throws {
        let (connection, channel) = try await self.makeTestConnectionWithAsyncTestingChannel()

        connection.addListener(channel: "example") { context, notification in
            XCTFail("Did not expect to receive")
        }

        try await connection.close()

        XCTAssertEqual(channel.isActive, false)
    }
```
<kbd>
<img width="1268" alt="Screenshot 2024-06-23 at 1 16 11 AM" src="https://github.com/vapor/postgres-nio/assets/54685446/9772a138-dd83-46ff-adcb-aae113a1625e">
</kbd>